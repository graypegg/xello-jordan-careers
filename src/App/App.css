.App__header {
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 1.8rem;
  -webkit-font-smoothing: antialiased;
  background: var(--primary-gradient);
  color: #FFF;
  padding: 30px 40px 26px;
}

.App__header > img {
  margin-top: -4px;
  margin-right: 15px;
  width: 93px;
}

.App__sidebar--open,
.App__sidebar--closed {
  position: fixed;
  left: 0;
  top: 0;
  background: var(--dark);
  height: 100%;
  padding: 25px 30px;
  color: #FFF;
  font-weight: normal;
  font-size: 0.9rem;
  min-width: 220px;
  box-shadow: 0 0 20px rgba(20, 30, 20, 0.3);
  transition: transform 0.5s cubic-bezier(0.19, 1, 0.22, 1);;
}

.App__sidebar--open {
  transform: translateX(-100%);
}

.App__sidebar--closed {
  transform: translateX(0);
}

.App__sidebarToggleButton {
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;
  left: 100%;
  top: 105px;
  height: 50px;
  width: 50px;
  border-top-right-radius: 10px;
  border-bottom-right-radius: 10px;
  background: var(--dark);
  cursor: pointer;
}

.App__sidebarToggleButton:hover {
  background: var(--darker);
}

.App__application {
  margin: 15px auto 0;
  max-width: 970px;
  padding: 0 40px; 
}