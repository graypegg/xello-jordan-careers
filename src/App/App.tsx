import { ICareer, IControlsState, IBookmark } from '../types'

import * as React from 'react'
import Sidebar from './Sidebar/Sidebar'
import BookmarkList from './BookmarkList/BookmarkList'
import CareerList from './CareerList/CareerList'
import Controls from './Controls/Controls'

import './App.css'
import iconBookmark from '../assets/images/icon-bookmark.svg'
import iconLogo from '../assets/images/logo.svg'
import { EStatus } from '../consts'
import { getLatest, postSave } from '../apiService/api.service'
import { isEqual as _isEqual } from 'lodash'

interface IAppProps {
  careers: ICareer[],
  offline?: boolean | null
}

interface IAppState {
  controlsState: IControlsState,
  bookmarks: IBookmark[],
  sidebarOpen: boolean,
  careersMeta: {[careerId: number]: ICareer['meta']},
  isDirty: boolean
}

class App extends React.Component<IAppProps, IAppState> {
  public showImages: boolean = true;
  public state: IAppState = {
    controlsState: {
      searchString: '',
      showImages: false,
      showStatuses: ((Object as any).values(EStatus)) as EStatus[],
      currentRevision: null
    },
    bookmarks: JSON.parse(localStorage.getItem('bookmarks') || '[]'),
    sidebarOpen: false,
    careersMeta: JSON.parse(localStorage.getItem('careersMeta') || '{}'),
    isDirty: false
  }

  constructor (props: IAppProps) {
    super(props)

    this.onControlsStateChange = this.onControlsStateChange.bind(this)
    this.onSaveBookmark = this.onSaveBookmark.bind(this)
    this.onDeleteBookmark = this.onDeleteBookmark.bind(this)
    this.updateBookmarks = this.updateBookmarks.bind(this)
    this.updateCareersMeta = this.updateCareersMeta.bind(this)
    this.onChangeSingleCareer = this.onChangeSingleCareer.bind(this)
    this.restore = this.restore.bind(this)
    this.save = this.save.bind(this)
  }

  public componentDidMount () {
    if (!this.props.offline) {
      getLatest()
        .then((latest) => {
          if (
            !(_isEqual(latest.data, {
              bookmarks: this.state.bookmarks,
              careersMeta: this.state.careersMeta
            }))
          ) {
            this.setState({
              isDirty: true
            })
          } else {
            this.setState((prevState) => ({
              controlsState: {
                ...prevState.controlsState,
                currentRevision: latest.id
              }
            }))
          }
        })
        .catch(() => {
          console.warn('You\'re offline. Can\'t check if we\'re working with the latest copy of the db or not')
        })
    }
  }

  public onControlsStateChange (controlsState: IControlsState): void {
    this.setState({ controlsState })
  }

  public filterCareers (careers: ICareer[], controlsState: IControlsState): ICareer[] {
    const searchStringRegExp = new RegExp(`(${controlsState.searchString})`, 'ig')
    return careers
      .filter((career) => searchStringRegExp.test(career.title))
      .filter((career) => career.meta
        ? controlsState.showStatuses.indexOf(career.meta.status || EStatus.NotStarted) > -1
        : controlsState.showStatuses.indexOf(EStatus.NotStarted) > -1
      )
  }

  public onSaveBookmark (career: ICareer) {
    this.updateBookmarks(
      this.state.bookmarks.concat([
        {
          career,
          saved: new Date(Date.now())
        }
      ])
    )
  }

  public onDeleteBookmark (career: ICareer) {
    this.updateBookmarks(
      this.state.bookmarks.filter((bookmark) => {
        return bookmark.career.id !== career.id
      })
    )
  }

  public updateBookmarks (bookmarks: IBookmark[]) {
    this.setState({
      bookmarks,
      isDirty: true
    })

    localStorage.setItem('bookmarks', JSON.stringify(bookmarks))
  }
  
  public mergeWithCareersMeta (careers: ICareer[]): ICareer[] {
    return careers.map((career) => {
      return {
        ...career,
        meta: this.state.careersMeta[career.id]
      }
    })
  }

  public mergeBookmarksWithCareersMeta (bookmarks: IBookmark[], careersMeta: {[careerId: number]: ICareer['meta']} | null): IBookmark[] {
    return bookmarks.map((bookmark) => {
      return {
        ...bookmark,
        career: {
          ...bookmark.career,
          meta: careersMeta ? careersMeta[bookmark.career.id] : {}
        }
      }
    })
  }

  public updateCareersMeta (careers: ICareer[]): void {
    const careersMeta = careers.reduce((acc, career: ICareer) => {
      if (career.meta) {
        acc[career.id] = career.meta
      } else {
        acc[career.id] = null
      }
      return acc
    }, {})

    this.setState({
      careersMeta: Object.assign({}, this.state.careersMeta, careersMeta),
      isDirty: true
    })

    localStorage.setItem('careersMeta', JSON.stringify(Object.assign({}, this.state.careersMeta, careersMeta)))
  }

  public onChangeSingleCareer (career: ICareer): void {
    this.updateCareersMeta([career])
  }

  public restore (): Promise<any> {
    return getLatest()
      .then((obj) => {
        return new Promise((res) => {
          localStorage.setItem('bookmarks', JSON.stringify(obj.data.bookmarks))
          localStorage.setItem('careersMeta', JSON.stringify(obj.data.careersMeta))
          this.setState((prevState) => ({
            bookmarks: obj.data.bookmarks,
            careersMeta: obj.data.careersMeta,
            controlsState: {
              ...prevState.controlsState,
              currentRevision: obj.id
            },
            isDirty: false
          }), () => res())
        })
      })
  }

  public save (): Promise<any> {
    return postSave({
      bookmarks: this.state.bookmarks,
      careersMeta: this.state.careersMeta
    })
      .then((obj) => {
        return new Promise((res) => {
          this.setState((prevState) => ({
            controlsState: {
              ...prevState.controlsState,
              currentRevision: obj.id,
            },
            isDirty: false
          }), () => res())
        })
      })
  }

  public render(): JSX.Element {
    return (
      <div className="App__wrapper">
        <header className="App__header">
          <img src={iconLogo} alt="Xello" /> Careers
        </header>

        <Sidebar icon={iconBookmark}>
          <BookmarkList bookmarks={this.mergeBookmarksWithCareersMeta(this.state.bookmarks, this.state.careersMeta)} onChange={this.updateBookmarks} onChangeCareer={this.onChangeSingleCareer} />
        </Sidebar>

        <main className="App__application">
          <Controls
            stateIsDirty={this.state.isDirty}
            onChange={this.onControlsStateChange}
            controlsState={this.state.controlsState}
            onGlobalRestore={this.props.offline ? undefined : this.restore}
            onGlobalSave={this.props.offline ? undefined : this.save} />

          <CareerList
            careers={this.filterCareers(this.mergeWithCareersMeta(this.props.careers), this.state.controlsState)}
            bookmarks={this.state.bookmarks}
            showImages={this.state.controlsState.showImages}
            pageLength={30}
            onSaveBookmark={this.onSaveBookmark}
            onDeleteBookmark={this.onDeleteBookmark}
            onChangeCareers={this.updateCareersMeta} />
          </main>
      </div>
    )
  }
}

export default App
