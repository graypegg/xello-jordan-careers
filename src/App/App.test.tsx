import * as React from 'react'
import { shallow } from 'enzyme'
import App from './App'

it('renders without crashing', () => {
  shallow(<App careers={[]} />)
})

it('can update search state', () => {
  const component = shallow(<App careers={[]} />)
  expect(component.state('searchState').searchString).toBe('')

  const instance = component.instance() as App
  instance.onSearchStateChange({ searchString: 'test' })

  expect(component.state('searchState').searchString).toBe('test')
})