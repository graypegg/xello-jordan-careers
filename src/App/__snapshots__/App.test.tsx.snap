// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookmarks: can merge bookmark career data with meta data 1`] = `
Array [
  Object {
    "career": Object {
      "description": "Test",
      "id": 1,
      "image": "TestImage",
      "meta": Object {
        "status": "Complete",
      },
      "notes": Array [
        "a",
        "b",
      ],
      "title": "Test",
    },
    "saved": 1970-01-01T00:00:00.000Z,
  },
]
`;

exports[`bookmarks: can merge bookmark career data without meta data 1`] = `
Array [
  Object {
    "career": Object {
      "description": "Test",
      "id": 1,
      "image": "TestImage",
      "meta": Object {},
      "notes": Array [
        "a",
        "b",
      ],
      "title": "Test",
    },
    "saved": 1970-01-01T00:00:00.000Z,
  },
]
`;

exports[`bookmarks: can save new bookmark 1`] = `Array []`;

exports[`bookmarks: can update bookmark state in it's entirity 1`] = `Array []`;

exports[`bookmarks: can update bookmark state in it's entirity 2`] = `
Array [
  Object {
    "career": Object {
      "description": "Test",
      "id": 1,
      "image": "TestImage",
      "notes": Array [
        "a",
        "b",
      ],
      "title": "Test",
    },
    "saved": 1970-01-01T00:00:00.000Z,
  },
]
`;

exports[`bookmarks: can update bookmark state in it's entirity 3`] = `
Array [
  Array [
    "bookmarks",
    "[{\\"career\\":{\\"title\\":\\"Test\\",\\"description\\":\\"Test\\",\\"notes\\":[\\"a\\",\\"b\\"],\\"image\\":\\"TestImage\\",\\"id\\":1},\\"saved\\":\\"1970-01-01T00:00:00.000Z\\"}]",
  ],
]
`;

exports[`bookmarks: dirting state after saving a bookmark 1`] = `false`;

exports[`bookmarks: dirting state after saving a bookmark 2`] = `true`;

exports[`career meta data: can merge career data with meta data 1`] = `
Array [
  Object {
    "description": "Test",
    "id": 1,
    "image": "TestImage",
    "meta": Object {
      "status": "Not Started",
    },
    "notes": Array [
      "a",
      "b",
    ],
    "title": "Test",
  },
]
`;

exports[`career meta data: can restore from server 1`] = `
Object {
  "1": Object {
    "status": "Complete",
  },
}
`;

exports[`career meta data: can restore from server 2`] = `
Array [
  Object {
    "career": Object {
      "description": "Test",
      "id": 1,
      "image": "TestImage",
      "notes": Array [
        "a",
        "b",
      ],
      "title": "Test",
    },
    "saved": "Wed Dec 31 1969 19:00:00 GMT-0500 (EST)",
  },
]
`;

exports[`career meta data: can restore from server 3`] = `
Object {
  "currentRevision": 1,
  "searchString": "",
  "showImages": false,
  "showStatuses": Array [
    "Not Started",
    "In Progress",
    "Complete",
  ],
}
`;

exports[`career meta data: can restore from server 4`] = `
Array [
  Array [
    "bookmarks",
    "[{\\"career\\":{\\"title\\":\\"Test\\",\\"description\\":\\"Test\\",\\"notes\\":[\\"a\\",\\"b\\"],\\"image\\":\\"TestImage\\",\\"id\\":1},\\"saved\\":\\"Wed Dec 31 1969 19:00:00 GMT-0500 (EST)\\"}]",
  ],
  Array [
    "careersMeta",
    "{\\"1\\":{\\"status\\":\\"Complete\\"}}",
  ],
]
`;

exports[`career meta data: can save to server 1`] = `
Object {
  "currentRevision": 1,
  "searchString": "",
  "showImages": false,
  "showStatuses": Array [
    "Not Started",
    "In Progress",
    "Complete",
  ],
}
`;

exports[`career meta data: can update careerMeta state 1`] = `
Object {
  "1": Object {
    "status": "Not Started",
  },
}
`;

exports[`career meta data: can update careerMeta state 2`] = `
Array [
  Array [
    "careersMeta",
    "{\\"1\\":{\\"status\\":\\"Not Started\\"}}",
  ],
]
`;

exports[`career meta data: can update careerMeta state with null data 1`] = `
Object {
  "1": null,
}
`;

exports[`career meta data: can update single record in careerMeta state 1`] = `
Object {
  "1": Object {
    "status": "Complete",
  },
}
`;

exports[`career meta data: dirting state after updating meta data 1`] = `false`;

exports[`career meta data: dirting state after updating meta data 2`] = `true`;

exports[`control state: can update control state 1`] = `
Object {
  "currentRevision": null,
  "searchString": "",
  "showImages": false,
  "showStatuses": Array [
    "Not Started",
    "In Progress",
    "Complete",
  ],
}
`;

exports[`control state: can update control state 2`] = `
Object {
  "currentRevision": 0,
  "searchString": "test",
  "showImages": false,
  "showStatuses": Array [],
}
`;
